spirv_glsl.cpp

```diff
@@ -3221,6 +3239,199 @@ string CompilerGLSL::access_chain(uint32_t base, const uint32_t *indices, uint32
+	{
+		SPIRV_CROSS_THROW("Access chains that result in an array can not be flattened");
+	}
+	else if (target_type.basetype == SPIRType::Struct)
+	{
+		flattened_access_chain_struct(expr, base, indices, count, target_type, offset);
+	}
+	else if (target_type.columns > 1)
+	{
+		flattened_access_chain_matrix(expr, base, indices, count, target_type, offset);
+	}
+	else
+	{
+		flattened_access_chain_vector_scalar(expr, base, indices, count, target_type, offset);
+	}
+}
+
+void CompilerGLSL::flattened_access_chain_struct(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (size_t i = 0; i < target_type.member_types.size(); ++i)
+	{
+		if (i != 0)
+			expr += ",";
+
+		const SPIRType &member_type = get<SPIRType>(target_type.member_types[i]);
+		uint32_t member_offset = type_struct_member_offset(target_type, uint32_t(i));
+
+		flattened_access_chain(expr, base, indices, count, member_type, offset + member_offset);
+	}
+
+	expr += ")";
+}
+
+void CompilerGLSL::flattened_access_chain_matrix(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	expr += type_to_glsl(target_type);
+	expr += "(";
+
+	for (uint32_t i = 0; i < target_type.columns; ++i)
+	{
+		if (i != 0)
+			expr += ",";
+
+		flattened_access_chain_vector_scalar(expr, base, indices, count, target_type, offset + i * 16);
+	}
+
+	expr += ")";
+}
+
+void CompilerGLSL::flattened_access_chain_vector_scalar(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, const SPIRType &target_type, uint32_t offset)
+{
+	if (target_type.basetype != SPIRType::Float)
+		SPIRV_CROSS_THROW("Access chains that use non-floating-point base types can not be flattened");
+
+	auto buffer_name = to_name(expression_type(base).self);
+
+	expr += buffer_name;
+	expr += "[";
+
+	uint32_t result = flattened_access_chain_offset(expr, base, indices, count, offset);
+	assert(result % 4 == 0);
+
+	uint32_t index = result / 4;
+
+	expr += convert_to_string(index / 4);
+	expr += "]";
+
+	static const char* swizzle[4][4] =
+	{
+		{ ".x", ".y", ".z", ".w" },
+		{ ".xy", ".yz", ".zw" },
+		{ ".xyz", ".yzw" },
+		{ "" }
+	};
+
+	assert(target_type.vecsize >= 1 && target_type.vecsize <= 4);
+	assert(swizzle[target_type.vecsize - 1][index % 4]);
+
+	expr += swizzle[target_type.vecsize - 1][index % 4];
+}
+
+uint32_t CompilerGLSL::flattened_access_chain_offset(std::string &expr, uint32_t base, const uint32_t *indices, uint32_t count, uint32_t offset)
+{
+	const auto *type = &expression_type(base);
+	uint32_t type_size = 0;
+
+	// For resolving array accesses, etc, keep a local copy for poking.
+	SPIRType temp;
+
+	for (uint32_t i = 0; i < count; i++)
+	{
+		uint32_t index = indices[i];
+
+		// Arrays
+		if (!type->array.empty())
+		{
+			// We have to modify the type, so keep a local copy.
+			if (&temp != type)
```

Comment:
Ok, so instead of this hack, it would probably make sense to save the type of the "previous" array type in OpTypeArray to fix this hack.

```
type = get<SPIRType>(type->previous_array_type);
```
or something similar.